account-compression/programs/spl-compression/src/lib.rs

```diff
@@ -0,0 +1,550 @@
+//!
+//! A buffer of proof-like changelogs is stored on-chain that allow multiple proof-based writes to succeed within the same slot.
+//! This is accomplished by fast-forwarding out-of-date (or possibly invalid) proofs based on information stored in the changelogs.
+//! See a copy of the whitepaper [here](https://drive.google.com/file/d/1BOpa5OFmara50fTvL0VIVYjtg-qzHCVc/view)
+//!
+//! While SPL ConcurrentMerkleTrees can generically store arbitrary information,
+//! one exemplified use-case is the [Bubblegum](https://github.com/metaplex-foundation/metaplex-program-library/tree/master/bubblegum) contract,
+//! which uses SPL-Compression to store encoded information about NFTs.
+//! The use of SPL-Compression within Bubblegum allows for:
+//! - up to 1 billion NFTs to be stored in a single account on-chain (>10,000x decrease in on-chain cost)
+//! - up to 2048 concurrent updates per slot
+//!
+//! Operationally, SPL ConcurrentMerkleTrees **must** be supplemented by off-chain indexers to cache information
+//! about leafs and to power an API that can supply up-to-date proofs to allow updates to the tree.
+//! All modifications to SPL ConcurrentMerkleTrees are settled on the Solana ledger via instructions against the SPL Compression contract.
+//! A production-ready indexer (Plerkle) can be found in the [Metaplex program library](https://github.com/metaplex-foundation/digital-asset-validator-plugin)
+
+use anchor_lang::{
+    emit,
+    prelude::*,
+    solana_program::sysvar::{clock::Clock, rent::Rent},
+};
+use borsh::{BorshDeserialize, BorshSerialize};
+use bytemuck::cast_slice_mut;
+use spl_concurrent_merkle_tree::node::{empty_node_cached, EMPTY};
+use std::mem::size_of;
+
+pub mod data_wrapper;
+pub mod error;
+pub mod events;
+pub mod state;
+pub mod zero_copy;
+
+use crate::data_wrapper::{wrap_event, Wrapper};
+use crate::error::AccountCompressionError;
+use crate::events::ChangeLogEvent;
+use crate::state::ConcurrentMerkleTreeHeader;
+use crate::zero_copy::ZeroCopy;
+
+/// Exported for Anchor / Solita
+pub use spl_concurrent_merkle_tree::{
+    concurrent_merkle_tree::ConcurrentMerkleTree, error::ConcurrentMerkleTreeError, node::Node,
+};
+
+declare_id!("GRoLLzvxpxxu2PGNJMMeZPyMxjAUH9pKqxGXV9DGiceU");
+
+/// Context for initializing a new SPL ConcurrentMerkleTree
+#[derive(Accounts)]
+pub struct Initialize<'info> {
+    #[account(zero)]
+    /// CHECK: This account will be zeroed out, and the size will be validated
+    pub merkle_roll: UncheckedAccount<'info>,
+
+    /// Authority that validates the content of the trees.
+    /// Typically a program, e.g., the Bubblegum contract validates that leaves are valid NFTs.
+    pub authority: Signer<'info>,
+
+    /// Program used to emit changelogs as instruction data.
+    /// See `WRAPYChf58WFCnyjXKJHtrPgzKXgHp6MD9aVDqJBbGh`
+    pub candy_wrapper: Program<'info, Wrapper>,
+}
+
+/// Context for inserting, appending, or replacing a leaf in the tree
```

Comment:
Could you also document the use of remaining accounts?

Reply:
Added